sha/ta/sha_ta.c

```diff
@@ -0,0 +1,450 @@
+	TEE_Result res = TEE_ERROR_OUT_OF_MEMORY;
+	void *message = NULL;
+	size_t message_sz = 0;
+	uint32_t hmac_len = 0;
+	void *b2 = NULL;
+
+	const uint32_t exp_param_types =
+		TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				TEE_PARAM_TYPE_NONE,
+				TEE_PARAM_TYPE_NONE);
+
+	DMSG("Session %p: sha update operation", session);
+	sess = session;
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	message = params[0].memref.buffer;
+	message_sz = params[0].memref.size;
+	hmac_len = (uint32_t)params[1].memref.size;
+
+	if (params[1].memref.buffer && params[1].memref.size) {
+		b2 = TEE_Malloc(params[1].memref.size, 0);
+		if (!b2)
+			goto out;
+	}
+
+	TEE_MACUpdate(sess->op_handle, message, message_sz);
+
+	res = TEE_MACComputeFinal(sess->op_handle, message, message_sz,
+				  b2, &hmac_len);
+
+	if (res == TEE_SUCCESS)
+		TEE_MemMove(params[1].memref.buffer, b2, hmac_len);
+
+	params[1].memref.size = hmac_len;
+
+out:
+	TEE_Free(b2);
+	return res;
+}
+
+static TEE_Result compare_hmac_sha_algo(void *session, uint32_t param_types,
+					TEE_Param params[4])
+{
+	struct sha_hmac_algo *sess = NULL;
+	void *message = NULL;
+	size_t message_sz = 0;
+	void *hmac_buff = NULL;
+	uint32_t hmac_len = 0;
+
+	const uint32_t exp_param_types =
+		TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				TEE_PARAM_TYPE_MEMREF_INPUT,
+				TEE_PARAM_TYPE_NONE,
+				TEE_PARAM_TYPE_NONE);
+
+	DMSG("Session %p: Compare HMAC SHA", session);
+	sess = session;
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	message = params[0].memref.buffer;
+	message_sz = params[0].memref.size;
+	hmac_buff = params[1].memref.buffer;
+	hmac_len = (uint32_t)params[1].memref.size;
+
+	TEE_MACUpdate(sess->op_handle, message, message_sz);
+
+	return TEE_MACCompareFinal(sess->op_handle, message, message_sz,
+				   hmac_buff, hmac_len);
+}
+
+TEE_Result TA_CreateEntryPoint(void)
+{
+	/* Nothing to do */
+	return TEE_SUCCESS;
+}
+
+void TA_DestroyEntryPoint(void)
+{
+	/* Nothing to do */
+}
+
+TEE_Result TA_OpenSessionEntryPoint(uint32_t __unused param_types,
+					TEE_Param __unused params[4],
+					void **session)
+{
+	/*
+	 * Allocate and init for the session.
+	 * The address of the structure is used as session ID for
+	 * the client.
+	 */
+	struct sha_hmac_algo *sess = TEE_Malloc(sizeof(*sess), 0);
+
+	if (!sess)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sess->key_handle = TEE_HANDLE_NULL;
+	sess->op_handle = TEE_HANDLE_NULL;
+
+	*session = sess;
+	DMSG("Session %p: newly allocated", *session);
+
+	return TEE_SUCCESS;
+}
+
+void TA_CloseSessionEntryPoint(void *session)
+{
+	/* Get context from session ID */
+	DMSG("Session %p: release session", session);
+	struct sha_hmac_algo *sess = session;
+
+	/* Release the session resources */
+	if (sess->key_handle != TEE_HANDLE_NULL)
+		TEE_FreeTransientObject(sess->key_handle);
+	if (sess->op_handle != TEE_HANDLE_NULL)
+		TEE_FreeOperation(sess->op_handle);
```

Comment:
The functions support handles `TEE_HANDLE_NULL`:
```suggestion
	TEE_FreeTransientObject(sess->key_handle);
	TEE_FreeOperation(sess->op_handle);